src/shared_utilities/DataHandling.jl

```diff
@@ -0,0 +1,246 @@
+    SPACE <: ClimaCore.Spaces.AbstractSpace,
+    REF_DATE <: Dates.DateTime,
+    TSTART <: AbstractFloat,
+    DATES <: AbstractArray{<:Dates.DateTime},
+    DIMS,
+    TIMES <: AbstractArray{<:AbstractFloat},
+    CACHE,
+}
+    file_reader::FR
+    regridder::REG
+    target_space::SPACE
+    dimensions::DIMS
+    available_dates::DATES
+    t_start::TSTART
+    reference_date::REF_DATE
+    available_times::TIMES
+    _cached_regridded_fields::CACHE
+end
+
+"""
+    DataHandler(file_path::AbstractString,
+                varname::AbstractString,
+                target_space::ClimaCore.Spaces.AbstractSpace;
+                reference_date::Dates.DateTime = Dates.DateTime(1979, 1, 1),
+                t_start::AbstractFloat = 0.0,
+                regridder_type = :TempestRegridder)
+
+The DataHandler maintains a cache of Fields that were previously computed.
+
+TODO: Add function to clear cache, and/or CACHE_MAX_SIZE
+
+Positional arguments
+=====================
+
+- `file_path`: Path of the NetCDF file that contains the data
+
+- `varname`: Name of the dataset in the NetCDF that has to be read and processed
+
+- `target_space`: Space where the simulation is run, where the data has to be regridded to
+
+Keyword arguments
+===================
+
+Time/date information will be ignored for static input files. (They are still set to make
+everything more type stable.)
+
+- `reference_date`: Calendar date corresponding to the start of the simulation.
+
+- `t_start`: Simulation time at the beginning of the simulation. Typically this is 0
+             (seconds), but if might be different if the simulation was restarted.
+
+- `regridder_type`: What type of regridding to perform. Currently, the only one implemented
+                    is `:tempest` to use `TempestRemap`. `TempestRemap` regrids everything
+                    ahead of time and saves the result to HDF5 files.
+"""
+function DataHandler(
+    file_path::AbstractString,
+    varname::AbstractString,
+    target_space::ClimaCore.Spaces.AbstractSpace;
+    reference_date::Dates.DateTime = Dates.DateTime(1979, 1, 1),
+    t_start::AbstractFloat = 0.0,
+    regridder_type = :TempestRegridder,
+)
+
+    # File reader, deals with ingesting data, possibly buffered/cached
+    file_reader = NCFileReader(file_path, varname)
+
+    # Regridder, deals with converting and interpolating the input data onto the simulation
+    # grid
+    regridder_type in AVAILABLE_REGRIDDERS ||
+        error("Regridder $regridder_type not implemented")
+
+    regridder_args = Dict(
+        :TempestRegridder =>
+            (target_space, mktempdir(), varname, file_path),
+    )
+    RegridderConstructor = getfield(Regridder, regridder_type)
+    regridder = RegridderConstructor(regridder_args[regridder_type]...)
+
+    # NOTE: this is not concretely typed
+    _cached_regridded_fields = Dict{Dates.DateTime, ClimaCore.Fields.Field}()
+
+    available_dates = file_reader.available_dates
```

Comment:
available_dates and available_times do not have the same reference, that might be confusing? perhaps could document that (if i understand correctly) available_dates = list of dates in the file where there is data while available_times are the corresponding times of those dates in seconds since the start of the simulation?

Reply:
ah ok, it is described below! all good.